src/osemgrep/targeting/Include_filter.ml

```diff
@@ -0,0 +1,89 @@
+(*
+   Similar to Gitignore_filter but select paths to be kept rather than ignored.
+*)
+
+open Git_path.Ops
+
+type t = {
+  project_root : Fpath.t;
+  glob_matchers : Glob_matcher.t list;
+  no_match_loc : Glob_matcher.loc;
+}
+
+let check_nonnegated_pattern str =
+  match Gitignore_syntax.remove_negator str with
+  | None -> ()
+  | Some _ -> failwith ("--include patterns cannot be negated: " ^ str)
+
+let create ~project_root patterns =
+  List.iter check_nonnegated_pattern patterns;
+  let glob_matchers =
+    Common.map
+      (fun pat ->
+        Gitignore_syntax.parse_pattern
```

Comment:
Why not using the Glob_matcher parser instead? That way you don't even need to check
for negated patterns since Glob_matcher does not allow them anyway.

Reply:
It's because we need to add some anchoring depending on whether the pattern is considered anchored by git. Remember `a/` is unanchored but `a/b` is anchored. That nonsense is encoded in the `Gitignore_syntax` module.

The `check_nonnegated_pattern` call is to produce an error message if a pattern starts with an exclamation mark because it is almost certainly an error from the user. The glob pattern parser would otherwise accept `!` as a normal part of a path segment.